File: cpp/tests/test_replay.cpp
// ReplayEngine unit tests (API_CORE.md section 5).

#include <gtest/gtest.h>

#include <cstdint>
#include <vector>

#include "iap/marketdata/events.hpp"
#include "iap/replay/replay.hpp"

using iap::EventType;
using iap::MarketEvent;
using iap::ReplayEngine;

namespace {

constexpr std::uint8_t ADD = static_cast<std::uint8_t>(EventType::ADD);
constexpr std::uint8_t TRADE = static_cast<std::uint8_t>(EventType::TRADE);

// Small multi-instrument, multi-venue stream (event-time ordered).
std::vector<MarketEvent> sample_stream(int n) {
    std::vector<MarketEvent> out;
    std::uint64_t eid = 0;
    std::uint64_t seq[2][2] = {{0, 0}, {0, 0}};
    for (int i = 0; i < n; ++i) {
        std::uint32_t iid = (i % 3 == 0) ? 2u : 1u;
        std::uint16_t vid = (i % 2 == 0) ? 1u : 2u;
        std::uint64_t s = ++seq[iid - 1][vid - 1];
        ++eid;
        std::int64_t ts = 1000 + i;
        if (i % 5 == 4) {
            out.push_back(MarketEvent::of(eid, iid, vid, ts, ts + 10, s, TRADE,
                                          static_cast<std::uint8_t>(i % 2),
                                          100, 10 + i, 0, eid));
        } else {
            out.push_back(MarketEvent::of(eid, iid, vid, ts, ts + 10, s, ADD,
                                          static_cast<std::uint8_t>(i % 2),
                                          100 + (i % 7) - 3, 10 + i, 1000 + eid,
                                          0));
        }
    }
    return out;
}

}  // namespace

TEST(Replay, RoutesByInstrumentAndVenue) {
    ReplayEngine engine;
    auto events = sample_stream(30);
    auto summary = engine.run(events);
    EXPECT_EQ(summary.events_processed, 30u);
    EXPECT_EQ(summary.instruments, 2u);
    EXPECT_EQ(summary.time_regressions, 0u);
    const auto& books = engine.books();
    ASSERT_TRUE(books.count(1));
    ASSERT_TRUE(books.count(2));
    EXPECT_EQ(books.at(1).books().size(), 2u);  // venues 1 and 2
}

TEST(Replay, SnapshotEveryEmitsWithIndex) {
    ReplayEngine engine(0, 10);
    auto events = sample_stream(35);
    std::vector<std::uint64_t> seen;
    auto summary = engine.run(
        events, [&](std::uint64_t idx, const iap::ReplaySnapshot& snap) {
            seen.push_back(idx);
            EXPECT_EQ(snap.index, idx);
            EXPECT_FALSE(snap.states.empty());
        });
    EXPECT_EQ(summary.snapshots, 3u);
    ASSERT_EQ(engine.snapshots().size(), 3u);
    EXPECT_EQ(seen, (std::vector<std::uint64_t>{10, 20, 30}));
    EXPECT_EQ(engine.snapshots()[2].index, 30u);
}

TEST(Replay, CheckpointEveryKeepsLatestN) {
    ReplayEngine engine(5, 0, 3);
    auto events = sample_stream(40);
    engine.run(events);
    ASSERT_EQ(engine.checkpoints().size(), 3u);  // 8 taken, latest 3 kept
    EXPECT_EQ(engine.checkpoints()[0].events_processed, 30u);
    EXPECT_EQ(engine.checkpoints()[1].events_processed, 35u);
    EXPECT_EQ(engine.checkpoints()[2].events_processed, 40u);
}

TEST(Replay, TimeRegressionsCounted) {
    ReplayEngine engine;
    auto events = sample_stream(4);
    events[2].exchange_ts = events[1].exchange_ts - 100;  // regression
    events[2].receive_ts = events[2].exchange_ts + 1;
    engine.run(events);
    EXPECT_EQ(engine.time_regressions(), 1u);
}

TEST(Replay, DeterministicAcrossRuns) {
    auto events = sample_stream(60);
    ReplayEngine a, b;
    a.run(events);
    b.run(events);
    EXPECT_EQ(a.book_states(), b.book_states());
    EXPECT_EQ(a.checkpoint(), b.checkpoint());
}

TEST(Replay, CheckpointRestartEquivalence) {
    auto events = sample_stream(50);
    // One-pass run.
    ReplayEngine full;
    full.run(events);
    // Prefix -> checkpoint -> restore -> remainder.
    ReplayEngine prefix;
    std::vector<MarketEvent> head(events.begin(), events.begin() + 23);
    std::vector<MarketEvent> tail(events.begin() + 23, events.end());
    prefix.run(head);
    ReplayEngine resumed = ReplayEngine::restore(prefix.checkpoint());
    resumed.run(tail);
    EXPECT_EQ(resumed.events_processed(), full.events_processed());
    EXPECT_EQ(resumed.book_states(), full.book_states());
    EXPECT_EQ(resumed.checkpoint(), full.checkpoint());
}

TEST(Replay, RestorePreservesConfigAndCounters) {
    ReplayEngine engine(7, 9);
    auto events = sample_stream(20);
    engine.run(events);
    auto cp = engine.checkpoint();
    EXPECT_EQ(cp.checkpoint_every, 7u);
    EXPECT_EQ(cp.snapshot_every, 9u);
    ReplayEngine restored = ReplayEngine::restore(cp);
    EXPECT_EQ(restored.events_processed(), engine.events_processed());
    EXPECT_EQ(restored.checkpoint(), cp);
}

TEST(Replay, BookStatesSortedAndComplete) {
    ReplayEngine engine;
    engine.run(sample_stream(30));
    auto states = engine.book_states();
    ASSERT_EQ(states.size(), 2u);
    auto it = states.begin();
    EXPECT_EQ(it->first, 1u);  // instrument keys ascending
    ++it;
    EXPECT_EQ(it->first, 2u);
    for (const auto& [iid, venues] : states) {
        (void)iid;
        EXPECT_FALSE(venues.empty());
    }
}

File: python/tests/test_replay.py
"""Deterministic replay + checkpoint/restart equivalence tests."""

import json

import pytest

from iap.core.codec import read_jsonl
from iap.replay.replay import ReplayEngine


@pytest.fixture(scope="module")
def eq_events(golden_dir):
    return read_jsonl(golden_dir / "events_eq_mbo.jsonl")


@pytest.fixture(scope="module")
def fx_events(golden_dir):
    return read_jsonl(golden_dir / "events_fx_quote.jsonl")


def test_replay_deterministic(eq_events):
    a, b = ReplayEngine(), ReplayEngine()
    ra = a.run(eq_events)
    rb = b.run(eq_events)
    assert ra == rb
    assert a.checkpoint() == b.checkpoint()
    assert a.book_states() == b.book_states()


def test_replay_stats(eq_events):
    engine = ReplayEngine()
    stats = engine.run(eq_events)
    assert stats["events_processed"] == 2000
    assert stats["instruments"] == 1
    assert stats["time_regressions"] == 0


@pytest.mark.parametrize("split_at", [1, 100, 1234, 1999])
def test_checkpoint_restart_identical_continuation(eq_events, split_at):
    full = ReplayEngine()
    full.run(eq_events)

    part = ReplayEngine()
    part.run(eq_events[:split_at])
    cp = part.checkpoint()
    cp_json = json.loads(json.dumps(cp))  # survives serialization round-trip
    resumed = ReplayEngine.restore(cp_json)
    resumed.run(eq_events[split_at:])

    assert resumed.events_processed == full.events_processed
    assert resumed.checkpoint() == full.checkpoint()
    assert resumed.book_states() == full.book_states()


def test_periodic_checkpoints_are_replayable(eq_events):
    engine = ReplayEngine(checkpoint_every=500, keep_checkpoints=10)
    engine.run(eq_events)
    assert [cp["events_processed"] for cp in engine.checkpoints] == [500, 1000, 1500, 2000]
    # Resume from the 1000-event checkpoint and reach identical final state.
    resumed = ReplayEngine.restore(engine.checkpoints[1])
    resumed.run(eq_events[1000:])
    assert resumed.checkpoint() == engine.checkpoint()


def test_restore_carries_keep_checkpoints(eq_events):
    engine = ReplayEngine(checkpoint_every=100, keep_checkpoints=2)
    engine.run(eq_events[:600])
    assert [cp["events_processed"] for cp in engine.checkpoints] == [500, 600]
    cp = json.loads(json.dumps(engine.checkpoint()))
    resumed = ReplayEngine.restore(cp)
    assert resumed.keep_checkpoints == 2
    # the restored engine keeps enforcing the SAME retention policy
    resumed.run(eq_events[600:1100])
    assert [c["events_processed"] for c in resumed.checkpoints] == [1000, 1100]


def test_restore_round_trips_resting_order_arrival_order(eq_events):
    """resting_orders() must iterate identically after checkpoint/restore."""
    from iap.orderbook.book import OrderBook

    book = OrderBook(1, 1)
    for ev in eq_events[:800]:
        book.apply(ev)
    restored = OrderBook.restore(json.loads(json.dumps(book.checkpoint())))
    assert restored.resting_orders() == book.resting_orders()
    assert restored.resting_orders(0) == book.resting_orders(0)
    assert restored.resting_orders(1) == book.resting_orders(1)


def test_snapshot_emission(eq_events):
    seen = []
    engine = ReplayEngine(snapshot_every=250)
    engine.run(eq_events, on_snapshot=lambda i, snap: seen.append(i))
    assert seen == list(range(250, 2001, 250))
    assert len(engine.snapshots) == 8
    snap = engine.snapshots[-1]
    assert snap["index"] == 2000
    state = snap["instruments"]["1"]["1"]
    assert state["sequence"] == 2000
    assert state["best_bid_ticks"] < state["best_ask_ticks"]


def test_replay_multi_venue_fx(fx_events):
    engine = ReplayEngine()
    stats = engine.run(fx_events)
    assert stats["events_processed"] == 800
    cons = engine.instrument_book(101)
    assert sorted(cons.books) == [10, 11, 12]
    for vid, book in cons.books.items():
        assert not book.stale
        assert book.duplicates_dropped == 0 and book.gaps_detected == 0
    # Consolidated best must be at least as good as any single venue's.
    bb = cons.best_bid()
    assert bb is not None
    assert bb[0] == max(b.best_bid()[0] for b in cons.books.values() if b.best_bid())


def test_replay_golden_states_match_expected(eq_events, golden_dir):
    with open(golden_dir / "expected_book_states.json") as f:
        expected = json.load(f)
    engine = ReplayEngine(snapshot_every=100)
    engine.run(eq_events)
    for idx, exp_state in expected["states"].items():
        snap = next(s for s in engine.snapshots if s["index"] == int(idx))
        assert snap["instruments"]["1"]["1"] == exp_state


def test_invalid_cadence_rejected():
    with pytest.raises(ValueError):
        ReplayEngine(checkpoint_every=-1)
